src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+void *Esp32WiFiManager::wifi_manager_task(void *param)
+{
+    Esp32WiFiManager *wifi = static_cast<Esp32WiFiManager *>(param);
+
+    // Start the WiFi system before proceeding with remaining tasks.
+    wifi->start_wifi_system();
+
+    while (true)
+    {
+        EventBits_t bits = xEventGroupGetBits(wifi->wifiStatusEventGroup_);
+        if(bits & DHCP_GOTIP_BIT)
+        {
+            // If we do not have not an uplink connection force a config reload
+            // to start the connection process.
+            if(!wifi->uplink_)
+            {
+                wifi->configReloadRequested_ = true;
+            }
+        }
+        else
+        {
+            // Since we do not have an IP address we need to shutdown any
+            // active connections since they will be invalid until a new IP
+            // has been provisioned.
+            wifi->stop_hub();
+            wifi->stop_uplink();
+
+            // Make sure we don't try and reload configuration since we can't
+            // create outbound connections at this time.
+            wifi->configReloadRequested_ = false;
+        }
+
+        // Check if there are configuration changes to pick up.
+        if(wifi->configReloadRequested_)
+        {
+            // Since we are loading configuration data, shutdown the hub and
+            // uplink if created previously.
+            wifi->stop_hub();
+            wifi->stop_uplink();
+
+            if(CDI_READ_TRIMMED(wifi->cfg_.sleep, wifi->configFd_))
+            {
+                // When sleep is enabled this will trigger the WiFi system to
+                // only wake up every DTIM period to receive beacon updates.
+                // no data loss is expected for this setting but it does delay
+                // receiption until the DTIM period.
+                ESP_ERROR_CHECK(esp_wifi_set_ps(WIFI_PS_MIN_MODEM));
+            }
+            else
+            {
+                // When sleep is disabled the WiFi radio will always be active.
+                // This will increase power consumption of the ESP32 but it
+                // will result in a more reliable behavior when the ESP32 is
+                // connected to an always-on power supply (ie: not a battery).
+                ESP_ERROR_CHECK(esp_wifi_set_ps(WIFI_PS_NONE));
+            }
+
+            if(CDI_READ_TRIMMED(wifi->cfg_.hub().enable, wifi->configFd_))
+            {
+                // Since hub mode is enabled start the HUB creation process.
+                wifi->start_hub();
+            }
+            // Start the uplink connection process in the background.
+            wifi->start_uplink();
+            wifi->configReloadRequested_ = false;
+        }
+
+        // Sleep until we are woken up again for configuration update or WiFi
+        // event.
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+    }
+
+    return nullptr;
+}
+
+// Shuts down the hub listener (if enabled and running) for this node.
```

Comment:
///
consider copying this comment to the class header.
Please document that this is a blocking call.

Reply:
oops, I missed adding that, let me add those before this merges.